[onnx.export] Track new nodes added during _run_symbolic_function

gustavla opened the pull request:
This PR is part of an effort to speed up torch.onnx.export (#121422).

- This copies the shape and type from the node to the nodes that are produced by the export. However, for 1-to-N exports, which are very common, this doesn't make much sense and can give the graph in broken shape or type information. As far as I can tell, a shape inference pass is used to propagate the correct shape and type for all interemediate (and final) nodes.
- If there is a situation where this is necessary (shape inference turned off and only 1-to-1 ops are exported ??), perhaps this can be conditionally skipped. It does incur a quadratic cost. Another option is to set a global default for the metadata and
use that for all nodes that get created. Again, this meta data may not make sense for all ops and seems dangerous to do.
- Resolves (8) in #121422.

(partial fix of #121422)

gustavla committed:
Do not blindly propagate metadata

- This copies the shape and type from the node to the nodes that are
  produced by the export. However, for 1-to-N exports, which are
  very common, this doesn't make much sense and can give the graph in
  broken shape or type information. As far as I can tell, a shape
  inference pass is used to propagate the correct shape and type for all
  interemediate (and final) nodes.
- If there is a situation where this is necessary (shape inference
  turned off and only 1-to-1 ops are exported ??), perhaps this can
  be conditionally skipped. It does incur a quadratic cost.
- Resolves (8) in #121422.

gustavla committed:
Only copy input shape data

- Before, all of the inputs from the graph were copied over into
  inferred_shape_data. This includes all constants of all nodes,
  and grows with the graph size. This incurs a quadratic time
  complexity. Instead, we only copy the shapes we need.
- Resolves (7) in #121422.

BowenBao commented on torch/csrc/jit/passes/onnx.cpp:
```diff
@@ -448,14 +444,6 @@ void NodeToONNX(
-    // propagate metadata
-    for (auto node : g->nodes()) {
-      if (nodes_before.find(node) == nodes_before.end()) {
-        node->copyMetadata(n);
```
This was introduced in #45256 and the main motivation is to propagate "sourceRange" and "scope" information.

The node level copyMetadata does not copy shape and type

https://github.com/pytorch/pytorch/blob/b4c810491ea726df4d3857a1eb8d46e9edc212f5/torch/csrc/jit/ir/ir.h#L436-L443

Indeed this is introducing quadratic cost. If profiling also suggests this being a hotspot, maybe we should add extra state to make the "new nodes" added during `_run_symbolic_function` retrievable.

gustavla replied:
Now I remember it's the sourceRange and scope info and not the shape/type. I got it mixed up. OK, I'll look into adding a state to make all new nodes inherit this. It can be sort of a "set default sourceRange and scope".

gustavla replied:
I have now pushed an update where the new nodes are being tracked by the `GraphContext`. @BowenBao Let me know what you think.

gustavla committed:
Track new nodes

- Add a new_nodes in GraphContext and append all new nodes.
- The new_nodes object is passed into _run_symbolic_function
  and new values are collected in place.
- That way we can still correctly copy the meta data into the new nodes,
  without incurring the quadratic time complexity cost.

BowenBao commented on torch/onnx/_internal/jit_utils.py:
```diff
@@ -42,6 +42,7 @@ class GraphContext:
     original_node: _C.Node
     params_dict: Dict[str, "_C.IValue"]
     env: Dict[_C.Value, _C.Value]
+    new_nodes: List[_C.Node]
```
Add field default factory of empty list to make it backward compatible

gustavla replied:
Thanks, I've pushed a fix.

BowenBao commented on torch/onnx/_internal/jit_utils.py:
```diff
@@ -42,6 +42,7 @@ class GraphContext:
     original_node: _C.Node
     params_dict: Dict[str, "_C.IValue"]
     env: Dict[_C.Value, _C.Value]
+    new_nodes: List[_C.Node]
```
Please update docstring of class `GraphContext` above with new attribute and the motivation.

gustavla replied:
Fixed. Thanks!

gustavla committed:
Add default value and docstring to new_nodes

gustavla commented:
@BowenBao Let me know if there are any other concerns here, thanks!

BowenBao reviewed (approved):
LGTM, thanks for contribution!

gustavla commented:
@BowenBao Great, thanks!

The CI failures look completely unrelated:
```
.ci/pytorch/test.sh: line 215: ulimit: stack size: cannot modify limit: Operation not permitted
```
I can try to rebase and see if that resolves it.

gustavla commented:
That seems to have done the trick. Is this ready to go in then?

thiagocrepaldi commented:
@pytorchbot merge
